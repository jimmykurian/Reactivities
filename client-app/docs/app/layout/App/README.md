[**Client App Documentation v0.0.0**](../../../README.md) â€¢ **Docs**

***

[Client App Documentation v0.0.0](../../../README.md) / app/layout/App

# app/layout/App

## Index

### Functions

- [default](functions/default.md)
