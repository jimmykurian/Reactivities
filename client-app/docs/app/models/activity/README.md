[**Client App Documentation v0.0.0**](../../../README.md) â€¢ **Docs**

***

[Client App Documentation v0.0.0](../../../README.md) / app/models/activity

# app/models/activity

## Index

### Interfaces

- [Activity](interfaces/Activity.md)
