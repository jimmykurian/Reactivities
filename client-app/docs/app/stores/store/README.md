[**Client App Documentation v0.0.0**](../../../README.md) â€¢ **Docs**

***

[Client App Documentation v0.0.0](../../../README.md) / app/stores/store

# app/stores/store

## Index

### Interfaces

- [Store](interfaces/Store.md)

### Variables

- [StoreContext](variables/StoreContext.md)
- [store](variables/store.md)

### Functions

- [useStore](functions/useStore.md)
