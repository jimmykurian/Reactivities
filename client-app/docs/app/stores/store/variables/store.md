[**Client App Documentation v0.0.0**](../../../../README.md) â€¢ **Docs**

***

[Client App Documentation v0.0.0](../../../../README.md) / [app/stores/store](../README.md) / store

# Variable: store

> `const` **store**: [`Store`](../interfaces/Store.md)

## Constant

store

## Description

The store instance containing all the application's stores.

## Source

[src/app/stores/store.ts:25](https://github.com/jimmykurian/Reactivities/blob/5670213c338d2af4595c1ce87506b162bb3099b4/client-app/src/app/stores/store.ts#L25)
