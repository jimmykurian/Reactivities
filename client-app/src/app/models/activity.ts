/**
 * @author Jimmy Kurian
 * @name Activity
 */

/**
 * The Activity interface represents the structure of an activity object in the Reactivities application.
 * It includes various properties such as id, title, date, description, category, city, and venue.
 *
 * @interface
 *
 * @property {string} id - The unique identifier for the activity.
 * @property {string} title - The title of the activity.
 * @property {string} date - The date of the activity in ISO string format.
 * @property {string} description - A brief description of the activity.
 * @property {string} category - The category to which the activity belongs.
 * @property {string} city - The city where the activity takes place.
 * @property {string} venue - The specific venue of the activity within the city.
 */
export interface Activity {
  id: string;
  title: string;
  date: string;
  description: string;
  category: string;
  city: string;
  venue: string;
}
