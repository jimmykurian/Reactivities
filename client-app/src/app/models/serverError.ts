/**
 * @file ServerError.ts
 * @author Jimmy Kurian
 * @fileoverview This file defines the `ServerError` interface, which models the structure of server error responses in the Reactivities application.
 */

/**
 * The ServerError interface represents the structure of an error returned from the server.
 * It includes the HTTP status code, an error message, and optional detailed information about the error.
 *
 * @interface ServerError
 *
 * @property {number} statusCode - The HTTP status code associated with the error.
 * @property {string} message - A brief message describing the error.
 * @property {string} details - Additional details or stack trace information about the error (optional).
 */
export interface ServerError {
  statusCode: number;
  message: string;
  details: string;
}
