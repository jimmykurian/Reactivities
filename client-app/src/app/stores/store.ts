/**
 * @author Jimmy Kurian
 * @name Store
 */

import { createContext, useContext } from 'react';
import ActivityStore from './activityStore';

/**
 * Interface representing the structure of the store.
 *
 * @interface Store
 * @property {ActivityStore} activityStore - Instance of ActivityStore for managing activity-related state.
 */
interface Store {
  activityStore: ActivityStore;
}

/**
 * The store instance containing all the application's stores.
 *
 * @constant
 * @type {Store}
 */
export const store: Store = {
  activityStore: new ActivityStore(),
};

/**
 * The React context for the store.
 *
 * @constant
 * @type {React.Context<Store>}
 */
export const StoreContext = createContext(store);

/**
 * Custom hook to use the store in React components.
 *
 * @function
 * @returns {Store} The store instance.
 *
 * @example
 * ```tsx
 * import { useStore } from './store';
 *
 * const MyComponent = () => {
 *   const { activityStore } = useStore();
 *   return <div>{activityStore.title}</div>;
 * };
 * ```
 */
export function useStore(): Store {
  return useContext(StoreContext);
}
