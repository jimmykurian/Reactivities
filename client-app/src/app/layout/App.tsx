/**
 * @author Jimmy Kurian
 * @name App
 */

import { useEffect, useState } from 'react';
import { Container } from 'semantic-ui-react';
import { Activity } from '../models/activity';
import NavBar from './NavBar';
import ActivityDashboard from '../features/activities/dashboard/ActivityDashboard';
import { v4 as uuid } from 'uuid';
import agent from '../api/agent';
import LoadingComponent from './LoadingComponent';

/**
 * The App component serves as the root component for the React application.
 * It fetches activities from an API, maintains the activities state, and renders the NavBar and ActivityDashboard components.
 *
 * @component
 * @returns {JSX.Element} The rendered App component.
 *
 * @remarks
 * The App component uses the `useEffect` hook to fetch activities from the API when the component mounts.
 * The `useState` hook is used to manage the activities state, the selected activity state, the edit mode state, the loading state, and the submitting state.
 * The component includes the NavBar and ActivityDashboard components for displaying the navigation bar and list of activities, respectively.
 * The `handleSelectActivity`, `handleCancelSelectActivity`, `handleFormOpen`, `handleFormClose`, `handleCreateOrEditActivity`, and `handleDeleteActivity` functions are used to manage the selected activity and edit mode states, as well as creating, editing, and deleting activities.
 *
 * @example
 * ```tsx
 * import React from 'react';
 * import ReactDOM from 'react-dom/client';
 * import App from './App';
 *
 * ReactDOM.createRoot(document.getElementById('root')!).render(
 *   <React.StrictMode>
 *     <App />
 *   </React.StrictMode>,
 * );
 * ```
 */
function App(): JSX.Element {
  const [activities, setActivities] = useState<Activity[]>([]);
  const [selectedActivity, setSelectedActivity] = useState<
    Activity | undefined
  >(undefined);
  const [editMode, setEditMode] = useState(false);
  const [loading, setLoading] = useState(true);
  const [submitting, setSubmitting] = useState(false);

  useEffect(() => {
    agent.Activities.list().then((response) => {
      const formattedActivities: Activity[] = response.map((activity) => ({
        ...activity,
        date: activity.date.split('T')[0], // Format date to 'YYYY-MM-DD'
      }));
      setActivities(formattedActivities);
      setLoading(false);
    });
  }, []);

  /**
   * Handles selecting an activity by its ID.
   *
   * @param {string} id - The ID of the activity to select.
   */
  function handleSelectActivity(id: string): void {
    setSelectedActivity(activities.find((a) => a.id === id));
  }

  /**
   * Handles canceling the selection of an activity.
   */
  function handleCancelSelectActivity(): void {
    setSelectedActivity(undefined);
  }

  /**
   * Opens the form for creating or editing an activity.
   *
   * @param {string} [id] - The ID of the activity to edit, or undefined to create a new activity.
   */
  function handleFormOpen(id?: string): void {
    id ? handleSelectActivity(id) : handleCancelSelectActivity();
    setEditMode(true);
  }

  /**
   * Closes the form for creating or editing an activity.
   */
  function handleFormClose(): void {
    setEditMode(false);
  }

  /**
   * Handles creating or editing an activity.
   *
   * @param {Activity} activity - The activity object to create or edit.
   */
  function handleCreateOrEditActivity(activity: Activity): void {
    setSubmitting(true);
    if (activity.id) {
      agent.Activities.update(activity).then(() => {
        setActivities(
          activities.map((a) => (a.id === activity.id ? activity : a)),
        );
        setSelectedActivity(activity);
        setEditMode(false);
        setSubmitting(false);
      });
    } else {
      activity.id = uuid();
      agent.Activities.create(activity).then(() => {
        setActivities([...activities, activity]);
        setSelectedActivity(activity);
        setEditMode(false);
        setSubmitting(false);
      });
    }
  }

  /**
   * Handles deleting an activity.
   *
   * @param {string} id - The ID of the activity to delete.
   */
  function handleDeleteActivity(id: string): void {
    setSubmitting(true);
    agent.Activities.delete(id).then(() => {
      setActivities(activities.filter((a) => a.id !== id));
      setSubmitting(false);
    });
  }

  if (loading) return <LoadingComponent content="Loading app..." />;

  return (
    <>
      <NavBar openForm={handleFormOpen} />
      <Container style={{ marginTop: '7em' }}>
        <ActivityDashboard
          activities={activities}
          selectedActivity={selectedActivity}
          selectActivity={handleSelectActivity}
          cancelSelectActivity={handleCancelSelectActivity}
          editMode={editMode}
          openForm={handleFormOpen}
          closeForm={handleFormClose}
          createOrEdit={handleCreateOrEditActivity}
          deleteActivity={handleDeleteActivity}
          submitting={submitting}
        />
      </Container>
    </>
  );
}

export default App;
