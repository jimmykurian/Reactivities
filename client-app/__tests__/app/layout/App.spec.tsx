import { render, screen, waitFor } from '@testing-library/react';
import App from '../../../src/app/layout/App';
import { Activity } from '../../../src/app/models/activity';
import { useStore } from '../../../src/app/stores/store';
import { makeAutoObservable, runInAction } from 'mobx';
import { BrowserRouter as Router, Route, Routes } from 'react-router-dom';

// Mock the agent and store
jest.mock('../../../src/app/api/agent');
jest.mock('../../../src/app/stores/store', () => ({
  useStore: jest.fn(),
}));

// Mocking ActivityDashboard component
const MockActivityDashboard = () => (
  <div>
    <div>Static Test Activity 1</div>
    <div>Static Test Activity 2</div>
  </div>
);

describe('App', () => {
  const staticMockActivities: Activity[] = [
    {
      id: '1',
      title: 'Static Test Activity 1',
      date: '2024-06-12',
      description: 'Static Description 1',
      category: 'Static Category 1',
      city: 'Static City 1',
      venue: 'Static Venue 1',
    },
    {
      id: '2',
      title: 'Static Test Activity 2',
      date: '2024-07-12',
      description: 'Static Description 2',
      category: 'Static Category 2',
      city: 'Static City 2',
      venue: 'Static Venue 2',
    },
  ];

  class MockActivityStore {
    activities: Activity[] = [];
    selectedActivity: Activity | undefined = undefined;
    editMode = false;
    loading = false;
    loadingInitial = true;
    loadActivities = jest.fn().mockImplementation(async () => {
      runInAction(() => {
        this.activities = staticMockActivities;
        this.loadingInitial = false;
      });
    });

    get activitiesByDate() {
      return this.activities
        .slice()
        .sort((a, b) => Date.parse(a.date) - Date.parse(b.date));
    }

    constructor() {
      makeAutoObservable(this);
    }
  }

  const mockStore = {
    activityStore: new MockActivityStore(),
  };

  beforeEach(() => {
    (useStore as jest.Mock).mockReturnValue(mockStore);
    jest.clearAllMocks();
  });

  test('renders Reactivities heading', async () => {
    // Arrange
    render(
      <Router>
        <Routes>
          <Route path="/" element={<App />}>
            <Route path="/" element={<MockActivityDashboard />} />
          </Route>
        </Routes>
      </Router>,
    );

    // Act
    const headingElement = await waitFor(() =>
      screen.getByText(/Reactivities/i),
    );

    // Assert
    expect(headingElement).toBeInTheDocument();
  });

  test('matches snapshot', async () => {
    // Arrange
    const { asFragment } = render(
      <Router>
        <Routes>
          <Route path="/" element={<App />}>
            <Route path="/" element={<MockActivityDashboard />} />
          </Route>
        </Routes>
      </Router>,
    );

    // Act
    await waitFor(() => {
      expect(
        screen.getAllByText(/Static Test Activity/i).length,
      ).toBeGreaterThan(0);
    });

    // Assert
    expect(asFragment()).toMatchSnapshot();
  });
});
