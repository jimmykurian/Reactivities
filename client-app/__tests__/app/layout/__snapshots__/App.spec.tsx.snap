// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App matches snapshot 1`] = `
<DocumentFragment>
  <div
    class="ui inverted top fixed menu"
  >
    <div
      class="ui container"
    >
      <div
        class="header item"
      >
        <img
          alt="logo"
          src="/assets/logo.png"
          style="margin-right: 10px;"
        />
        Reactivities
      </div>
      <div
        class="item"
        role="menuitem"
      >
        Activities
      </div>
      <div
        class="item"
      >
        <button
          class="ui positive button"
        >
          Create Activity
        </button>
      </div>
    </div>
  </div>
  <div
    class="ui container"
    style="margin-top: 7em;"
  >
    <h2>
      Hello from MobX!
    </h2>
    <button
      class="ui positive button"
    >
      Add exclamation!
    </button>
    <div
      class="ui grid"
    >
      <div
        class="ten wide column"
      >
        <div
          class="ui list"
          role="list"
        >
          <div
            class="ui segment"
          >
            <div
              class="ui divided items"
            >
              <div
                class="item"
              >
                <div
                  class="content"
                >
                  <a
                    class="header"
                  >
                    Static Test Activity 1
                  </a>
                  <div
                    class="meta"
                  >
                    2024-06-12
                  </div>
                  <div
                    class="description"
                  >
                    <div>
                      Static Description 1
                    </div>
                    <div>
                      Static City 1, Static Venue 1
                    </div>
                  </div>
                  <div
                    class="extra"
                  >
                    <button
                      class="ui blue right floated button"
                    >
                      View
                    </button>
                    <button
                      class="ui red right floated button"
                      name="1"
                    >
                      Delete
                    </button>
                    <div
                      class="ui basic label"
                    >
                      Static Category 1
                    </div>
                  </div>
                </div>
              </div>
              <div
                class="item"
              >
                <div
                  class="content"
                >
                  <a
                    class="header"
                  >
                    Static Test Activity 2
                  </a>
                  <div
                    class="meta"
                  >
                    2024-07-12
                  </div>
                  <div
                    class="description"
                  >
                    <div>
                      Static Description 2
                    </div>
                    <div>
                      Static City 2, Static Venue 2
                    </div>
                  </div>
                  <div
                    class="extra"
                  >
                    <button
                      class="ui blue right floated button"
                    >
                      View
                    </button>
                    <button
                      class="ui red right floated button"
                      name="2"
                    >
                      Delete
                    </button>
                    <div
                      class="ui basic label"
                    >
                      Static Category 2
                    </div>
                  </div>
                </div>
              </div>
            </div>
          </div>
        </div>
      </div>
      <div
        class="six wide column"
      />
    </div>
  </div>
</DocumentFragment>
`;
